lib/NetPacksLib.cpp

```diff
@@ -676,8 +676,24 @@ DLL_LINKAGE void GiveHero::applyGs(CGameState *gs)
 
 DLL_LINKAGE void NewObject::applyGs(CGameState *gs)
 {
-	const TerrainTile &t = gs->map->getTile(pos);
-	ETerrainType terrainType = t.terType;
+	ETerrainType terrainType;
+
+	if(ID == Obj::BOAT && !gs->isInTheMap(pos)) //special handling for bug #3060 - pos outside map but visitablePos is not
+	{
+		CGObjectInstance testObject = CGObjectInstance();
+		testObject.pos = pos;
+		testObject.appearance = VLC->objtypeh->getHandlerFor(ID, subID)->getTemplates(ETerrainType::WATER).front();
+
+		const int3 previousXAxisTile = int3(pos.x - 1, pos.y, pos.z);
+		bool test1 = gs->isInTheMap(previousXAxisTile);
+		int3 test2 = testObject.visitablePos();
+		assert(gs->isInTheMap(previousXAxisTile) && (testObject.visitablePos() == previousXAxisTile)); 

```

Comment:
What will be the value of terraintype at the end? As far as I understood getTile was invoked to get this variable

Reply:
few lines below:
```
switch(ID)
	{
	case Obj::BOAT:
		o = new CGBoat();
		terrainType = ETerrainType::WATER;
...
```

Reply:
Technically you say here assert(gs->isInTheMap(testObject.visitablePos())) which looks more logical (objects visitable pos should be in the map) and does not need mystic x - 1.

Reply:
Somewhat true, I made stricter check which ensures visitable position is one tile away to make it more logical for this particular issue.